# Weeks 3+ — Working with the full release (~79M rows) without downloading 79M rows

[![Open In Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/Sarahnjunge/starter-notebooks/blob/main/notebooks/03_working_with_the_full_release.ipynb?flush_cache=true)

Notebooks 01–02 used the small starter CSV that ships with this repo. Your lane and capstone work
run on the **full pseudonymized warehouse release**: ~17 months of daily search performance for
~70 clients, plus a query-level table. It is hosted as Parquet on Hugging Face, and the trick of
this notebook is that you **never download or load the whole thing** — DuckDB reads only the
columns and partitions your SQL touches.

By the end you will have:
1. Connected DuckDB to the hosted release and listed every table.
2. Pulled a **feature table you designed** (aggregates per content item) into pandas.
3. Trained a quick scikit-learn model on features you built from 79M rows — on a free Colab CPU.

**Before you start (one-time, ~2 minutes):**
1. Create a free [Hugging Face account](https://huggingface.co/join).
2. Open the dataset page ([`FlyRank/internship-warehouse`](https://huggingface.co/datasets/FlyRank/internship-warehouse)) and **request access** (instant after you accept the data-use terms). **Accept the terms in your browser first — the token below 401s until access is granted (usually instant).**
3. Create a **read** token at [Settings → Access Tokens](https://huggingface.co/settings/tokens). **Never paste the token into a code cell** — your repo is public; use the `getpass` prompt below (or Colab's 🔑 Secrets panel).


### 1) Data contract — Content Decline Prediction

**1. What does one row mean?**  
One row represents one content item for one client at a particular reporting date in the search-performance panel.

**2. Which table(s) will I use?**  
I will use `fact_daily` for content-level daily search performance and `fact_query_90d` for historical query-level signals.

**3. What time window will I use?**  
I will develop and verify the contract using March 2026 as a mid-panel development month, using only information available before the prediction window.

**4. What will I predict?**  
I will predict whether a content item's search impressions decline by more than 20% in the future outcome window compared with its preceding historical window.

**5. What will I deliberately exclude?**  
I will exclude any feature calculated from the future outcome window, including future impressions, future clicks, or future ranking position, because these would not be knowable at the decision moment.

In [20]:
%pip -q install duckdb huggingface_hub


In [21]:
import os, getpass

# CI and power users set HF_TOKEN in the environment; everyone else gets the safe prompt.
HF_TOKEN = os.environ.get('HF_TOKEN') or getpass.getpass('Paste your Hugging Face READ token (hf_...): ')


Paste your Hugging Face READ token (hf_...): ··········


## 1. Connect DuckDB to the release

DuckDB speaks `hf://` natively. The secret below authenticates every query; after that the
release behaves like a set of local tables.


In [22]:
import duckdb

con = duckdb.connect()
con.execute(f"CREATE OR REPLACE SECRET hf (TYPE huggingface, TOKEN '{HF_TOKEN}')")

REL = 'hf://datasets/FlyRank/internship-warehouse'
TABLES = {
    'dim_clients':                f"read_parquet('{REL}/dim_clients.parquet')",
    'dim_content':                f"read_parquet('{REL}/dim_content.parquet')",
    'fact_daily':                 f"read_parquet('{REL}/fact_content_daily_performance/**/*.parquet')",
    'fact_daily_sample':          f"read_parquet('{REL}/fact_content_daily_performance_sample.parquet')",
    'fact_query_90d':             f"read_parquet('{REL}/fact_content_query_90d.parquet')",
}

for name, src in TABLES.items():
    n = con.sql(f'SELECT COUNT(*) FROM {src}').fetchone()[0]
    print(f'{name:22} {n:>12,} rows')


dim_clients                     104 rows
dim_content                 519,606 rows


FloatProgress(value=0.0, layout=Layout(width='auto'), style=ProgressStyle(bar_color='black'))

fact_daily               78,835,655 rows
fact_daily_sample        11,694,072 rows
fact_query_90d            2,414,248 rows


That count over the daily fact touched **Parquet metadata, not data** — it finished in seconds
even though the table has ~79M rows. That is the whole workflow: push the heavy lifting into
DuckDB SQL, bring only small results into pandas.

## 2. Know your panel before you model it

History depth **differs per client** (an *unbalanced panel*). `dim_clients` tells you exactly
what each client has — check it before designing any time window.


In [23]:
clients = con.sql(f"""
    SELECT client_hash_id, access_profile, gsc_data_start, ga4_data_start
    FROM {TABLES['dim_clients']}
    ORDER BY gsc_data_start NULLS LAST
""").df()

print('clients with 12+ months of GSC history:',
      (clients['gsc_data_start'] <= clients['gsc_data_start'].dropna().max() - __import__('pandas').Timedelta(days=365)).sum())
clients.head(10)


clients with 12+ months of GSC history: 4


,client_hash_id,access_profile,gsc_data_start,ga4_data_start
0,client_9958f0a7ae1df715,gsc_and_ga4,2025-01-27,2025-10-29
1,client_ff644d8251367cbb,gsc_and_ga4,2025-01-27,2025-10-29
2,client_73cda7b4e4f265ea,gsc_and_ga4,2025-02-11,2026-03-24
3,client_fef1a8f436438636,gsc_and_ga4,2025-03-11,2026-03-06
4,client_62f4a7e64f5e0096,gsc_only,2025-06-07,NaT
5,client_b10cb2997d0c7c86,gsc_and_ga4,2025-06-18,2025-11-15
6,client_65de48885f4ef01b,gsc_and_ga4,2025-06-21,2026-02-19
7,client_c182d11e4862a37d,gsc_and_ga4,2025-06-21,2026-02-20
8,client_3197e6291363b4db,gsc_and_ga4,2025-06-29,2025-11-09
9,client_625b6439094e23e4,gsc_and_ga4,2025-07-01,2026-02-19


## 3. Build features with SQL, not with RAM

The pattern for every lane: **aggregate per content item inside DuckDB**, then hand the small
result to pandas/sklearn. Here: momentum features from the last 60 days of the panel.

**This is the heaviest cell in the notebook — expect 2–6 minutes on Colab.** It downloads ~2 months of column data over the network (RAM stays tiny; that's the point). If it runs past ~10 minutes or errors with `HTTP 429`, re-run this section against `TABLES['fact_daily_sample']` and save the full table for your final pass.


In [24]:
features = con.sql(f"""
    WITH daily AS (
        SELECT
            client_hash_id,
            content_hash_id,
            report_date,
            gsc_impressions,
            gsc_clicks,
            gsc_avg_position
        FROM {TABLES['fact_daily']}
        WHERE report_date >= DATE '2025-12-01'
          AND report_date < DATE '2026-03-01'
          AND gsc_data_available IS TRUE
    ),

    aggregated AS (
        SELECT
            client_hash_id,
            content_hash_id,

            SUM(gsc_impressions) AS imp_prev90,

            SUM(gsc_clicks) AS clk_prev90,

            SUM(gsc_clicks) /
                NULLIF(SUM(gsc_impressions), 0) AS ctr_prev90,

            STDDEV(gsc_avg_position) AS position_volatility,

            AVG(
                CASE
                    WHEN DAYOFWEEK(report_date) IN (0, 6)
                    THEN 1.0
                    ELSE 0.0
                END
            ) AS weekend_share,

            AVG(
                CASE
                    WHEN report_date >= DATE '2026-01-15'
                    THEN gsc_avg_position
                END
            ) AS position_recent45,

            AVG(
                CASE
                    WHEN report_date < DATE '2026-01-15'
                    THEN gsc_avg_position
                END
            ) AS position_older45

        FROM daily

        GROUP BY
            client_hash_id,
            content_hash_id

        HAVING SUM(gsc_impressions) >= 300
    )

    SELECT
        *,
        position_recent45 - position_older45 AS position_change
    FROM aggregated
""").df()

print(f"Feature rows: {len(features):,}")
features.head()

FloatProgress(value=0.0, layout=Layout(width='auto'), style=ProgressStyle(bar_color='black'))

Feature rows: 73,695


,client_hash_id,content_hash_id,imp_prev90,clk_prev90,ctr_prev90,position_volatility,weekend_share,position_recent45,position_older45,position_change
0,client_62f4a7e64f5e0096,content_b0253354aca1c60a,695.0,4.0,0.005755,18.906048,0.268293,16.552961,21.346104,-4.793143
1,client_62f4a7e64f5e0096,content_f6de871098b8adb4,1360.0,6.0,0.004412,11.205985,0.277778,5.561400,11.053290,-5.491890
2,client_62f4a7e64f5e0096,content_db716144c0f73ee2,2410.0,2.0,0.000830,2.739026,0.277778,2.572374,2.768748,-0.196374
3,client_62f4a7e64f5e0096,content_5545f5bb9c82d3db,937.0,0.0,0.000000,14.345368,0.279070,2.084761,9.014697,-6.929936
4,client_62f4a7e64f5e0096,content_ca0a59a69df7548a,3307.0,2.0,0.000605,11.365884,0.269663,4.811922,10.549006,-5.737085


## 4. Add query-level signals

`fact_content_query_90d` describes **how a page earns its impressions**: across how many
distinct queries, how concentrated, how much sits in the rare/anonymized tail. One page ranking
for 40 queries is a different animal from one page ranking for 2.


In [25]:
qsignals = con.sql(f"""
    SELECT content_hash_id,
           ANY_VALUE(content_visible_query_count)     AS visible_queries,
           ANY_VALUE(rare_impressions_share)          AS rare_share,
           ANY_VALUE(anonymized_impressions_share)    AS anon_share,
           MAX(impressions_90d)                       AS top_query_impressions,
           SUM(impressions_90d)                       AS kept_impressions
    FROM {TABLES['fact_query_90d']}
    GROUP BY content_hash_id
""").df()

qsignals['top_query_share'] = qsignals['top_query_impressions'] / qsignals['kept_impressions']
data = features.merge(qsignals, on='content_hash_id', how='left')
print(f'joined: {len(data):,} rows')
data.head()


FloatProgress(value=0.0, layout=Layout(width='auto'), style=ProgressStyle(bar_color='black'))

joined: 73,695 rows


,client_hash_id,content_hash_id,imp_prev90,clk_prev90,ctr_prev90,position_volatility,weekend_share,position_recent45,position_older45,position_change,visible_queries,rare_share,anon_share,top_query_impressions,kept_impressions,top_query_share
0,client_62f4a7e64f5e0096,content_b0253354aca1c60a,695.0,4.0,0.005755,18.906048,0.268293,16.552961,21.346104,-4.793143,15.0,0.225225,0.436036,88.0,376.0,0.234043
1,client_62f4a7e64f5e0096,content_f6de871098b8adb4,1360.0,6.0,0.004412,11.205985,0.277778,5.561400,11.053290,-5.491890,1.0,0.217949,0.735043,11.0,11.0,1.000000
2,client_62f4a7e64f5e0096,content_db716144c0f73ee2,2410.0,2.0,0.000830,2.739026,0.277778,2.572374,2.768748,-0.196374,16.0,0.042455,0.625320,357.0,1299.0,0.274827
3,client_62f4a7e64f5e0096,content_5545f5bb9c82d3db,937.0,0.0,0.000000,14.345368,0.279070,2.084761,9.014697,-6.929936,4.0,0.467532,0.376623,54.0,108.0,0.500000
4,client_62f4a7e64f5e0096,content_ca0a59a69df7548a,3307.0,2.0,0.000605,11.365884,0.269663,4.811922,10.549006,-5.737085,11.0,0.275701,0.500935,73.0,239.0,0.305439


In [26]:
print("Current data columns:")
print(data.columns.tolist())

Current data columns:
['client_hash_id', 'content_hash_id', 'imp_prev90', 'clk_prev90', 'ctr_prev90', 'position_volatility', 'weekend_share', 'position_recent45', 'position_older45', 'position_change', 'visible_queries', 'rare_share', 'anon_share', 'top_query_impressions', 'kept_impressions', 'top_query_share']


In [27]:
# Get the March 2026 outcome for each content page

future_outcome = con.sql(f"""
    SELECT
        client_hash_id,
        content_hash_id,
        SUM(gsc_impressions) AS imp_future30
    FROM {TABLES['fact_daily']}
    WHERE report_date >= DATE '2026-03-01'
      AND report_date < DATE '2026-04-01'
      AND gsc_data_available IS TRUE
    GROUP BY
        client_hash_id,
        content_hash_id
""").df()

print(f"Future outcome rows: {len(future_outcome):,}")

# Join historical features with the March outcome
data = features.merge(
    future_outcome,
    on=["client_hash_id", "content_hash_id"],
    how="inner"
)

# Define decline:
# March daily impressions are at least 20% lower
# than the previous 90-day daily average
data["is_declining"] = (
    (data["imp_future30"] / 31)
    < 0.8 * (data["imp_prev90"] / 90)
).astype(int)

print(f"Model rows: {len(data):,}")
print(f"Declining pages: {data['is_declining'].sum():,}")
print(f"Declining rate: {data['is_declining'].mean():.3f}")

FloatProgress(value=0.0, layout=Layout(width='auto'), style=ProgressStyle(bar_color='black'))

Future outcome rows: 176,738
Model rows: 69,607
Declining pages: 11,926
Declining rate: 0.171


## 5. A first honest model

Same shape as notebook 02: define a label, hold out data, compare against a dumb baseline.
Label: *did impressions decline by more than 20% month-over-month?* — built only from columns
that exist **before** the window we predict. (Momentum features from the last 30 days predicting
a label defined on those same 30 days would be leakage — so here the features come from the
prev-30 window and query-mix, and the label from the last-30 outcome.)


In [28]:
from sklearn.ensemble import RandomForestClassifier
from sklearn.model_selection import GroupShuffleSplit
from sklearn.metrics import classification_report, balanced_accuracy_score

# Features available before the March outcome
feature_cols = [
    'imp_prev90',
    'clk_prev90',
    'ctr_prev90',
    'position_volatility',
    'weekend_share',
    'position_change'
]

# Keep rows with all required features
model_data = data.dropna(
    subset=feature_cols + ['is_declining']
).copy()

X = model_data[feature_cols]
y = model_data['is_declining']
groups = model_data['client_hash_id']

# Split by client so the same client cannot appear
# in both training and test sets
gss = GroupShuffleSplit(
    n_splits=1,
    test_size=0.25,
    random_state=42
)

train_idx, test_idx = next(
    gss.split(X, y, groups=groups)
)

X_train = X.iloc[train_idx]
X_test = X.iloc[test_idx]
y_train = y.iloc[train_idx]
y_test = y.iloc[test_idx]

# Train the model
model = RandomForestClassifier(
    n_estimators=200,
    random_state=42,
    n_jobs=-1
)

model.fit(X_train, y_train)

# Evaluate
pred = model.predict(X_test)

print(f"Training rows: {len(X_train):,}")
print(f"Test rows: {len(X_test):,}")
print()
print(
    f"Base rate (always predict majority): "
    f"{max(y_test.mean(), 1 - y_test.mean()):.3f}"
)
print()
print(classification_report(y_test, pred, digits=3))
print(
    f"Balanced accuracy: "
    f"{balanced_accuracy_score(y_test, pred):.4f}"
)

Training rows: 23,869
Test rows: 34,209

Base rate (always predict majority): 0.851

              precision    recall  f1-score   support

           0      0.855     0.939     0.895     29125
           1      0.203     0.089     0.124      5084

    accuracy                          0.813     34209
   macro avg      0.529     0.514     0.509     34209
weighted avg      0.758     0.813     0.780     34209

Balanced accuracy: 0.5140


In [29]:
import pandas as pd

importance = (
    pd.Series(model.feature_importances_, index=feature_cols)
    .sort_values(ascending=False)
)

print(importance)

position_volatility    0.234783
imp_prev90             0.213196
position_change        0.210712
ctr_prev90             0.168181
clk_prev90             0.095391
weekend_share          0.077737
dtype: float64


In [30]:
# Compare the 90-day model with and without the added position_change feature

baseline_features = [
    'imp_prev90',
    'clk_prev90',
    'ctr_prev90',
    'position_volatility',
    'weekend_share'
]

X_base = model_data[baseline_features]
y_base = model_data['is_declining']
groups_base = model_data['client_hash_id']

gss_base = GroupShuffleSplit(
    n_splits=1,
    test_size=0.25,
    random_state=42
)

train_idx, test_idx = next(
    gss_base.split(X_base, y_base, groups=groups_base)
)

X_train_base = X_base.iloc[train_idx]
X_test_base = X_base.iloc[test_idx]
y_train_base = y_base.iloc[train_idx]
y_test_base = y_base.iloc[test_idx]

baseline_model = RandomForestClassifier(
    n_estimators=200,
    random_state=42,
    n_jobs=-1
)

baseline_model.fit(X_train_base, y_train_base)

baseline_pred = baseline_model.predict(X_test_base)

print("90-day model WITHOUT position_change")
print(
    f"Balanced accuracy: "
    f"{balanced_accuracy_score(y_test_base, baseline_pred):.4f}"
)
print()
print(classification_report(y_test_base, baseline_pred, digits=3))

90-day model WITHOUT position_change
Balanced accuracy: 0.5218

              precision    recall  f1-score   support

           0      0.857     0.911     0.884     29125
           1      0.207     0.132     0.161      5084

    accuracy                          0.796     34209
   macro avg      0.532     0.522     0.522     34209
weighted avg      0.761     0.796     0.776     34209



In [31]:
# Rebuild the 90-day features with a stricter history threshold

features_500 = con.sql(f"""
    SELECT
        client_hash_id,
        content_hash_id,
        SUM(gsc_impressions) AS imp_prev90,
        SUM(gsc_clicks) AS clk_prev90,
        SUM(gsc_clicks) / NULLIF(SUM(gsc_impressions), 0) AS ctr_prev90,
        STDDEV(gsc_avg_position) AS position_volatility,
        AVG(
            CASE
                WHEN DAYOFWEEK(report_date) IN (0, 6)
                THEN 1.0
                ELSE 0.0
            END
        ) AS weekend_share,
        AVG(
            CASE
                WHEN report_date >= DATE '2026-01-15'
                THEN gsc_avg_position
            END
        ) AS position_recent45,
        AVG(
            CASE
                WHEN report_date < DATE '2026-01-15'
                THEN gsc_avg_position
            END
        ) AS position_older45
    FROM {TABLES['fact_daily']}
    WHERE report_date >= DATE '2025-12-01'
      AND report_date < DATE '2026-03-01'
      AND gsc_data_available IS TRUE
    GROUP BY
        client_hash_id,
        content_hash_id
    HAVING SUM(gsc_impressions) >= 500
""").df()

features_500["position_change"] = (
    features_500["position_recent45"]
    - features_500["position_older45"]
)

print(f"Rows with >=500 impressions: {len(features_500):,}")

FloatProgress(value=0.0, layout=Layout(width='auto'), style=ProgressStyle(bar_color='black'))

Rows with >=500 impressions: 63,019


In [32]:
# Join the 500-impression feature table to the March outcome

data_500 = features_500.merge(
    future_outcome,
    on=["client_hash_id", "content_hash_id"],
    how="inner"
)

data_500["is_declining"] = (
    (data_500["imp_future30"] / 31)
    < 0.8 * (data_500["imp_prev90"] / 90)
).astype(int)

print(f"Model rows: {len(data_500):,}")
print(f"Declining pages: {data_500['is_declining'].sum():,}")
print(f"Declining rate: {data_500['is_declining'].mean():.3f}")

Model rows: 59,975
Declining pages: 10,354
Declining rate: 0.173


In [33]:
# Train and evaluate the 500-impression threshold model

feature_cols_500 = [
    'imp_prev90',
    'clk_prev90',
    'ctr_prev90',
    'position_volatility',
    'weekend_share',
    'position_change'
]

model_data_500 = data_500.dropna(
    subset=feature_cols_500 + ['is_declining']
).copy()

X_500 = model_data_500[feature_cols_500]
y_500 = model_data_500['is_declining']
groups_500 = model_data_500['client_hash_id']

gss_500 = GroupShuffleSplit(
    n_splits=1,
    test_size=0.25,
    random_state=42
)

train_idx_500, test_idx_500 = next(
    gss_500.split(X_500, y_500, groups=groups_500)
)

X_train_500 = X_500.iloc[train_idx_500]
X_test_500 = X_500.iloc[test_idx_500]
y_train_500 = y_500.iloc[train_idx_500]
y_test_500 = y_500.iloc[test_idx_500]

model_500 = RandomForestClassifier(
    n_estimators=200,
    random_state=42,
    n_jobs=-1
)

model_500.fit(X_train_500, y_train_500)

pred_500 = model_500.predict(X_test_500)

print("90-day model with >=500 impressions")
print(
    f"Balanced accuracy: "
    f"{balanced_accuracy_score(y_test_500, pred_500):.4f}"
)
print()
print(classification_report(y_test_500, pred_500, digits=3))

90-day model with >=500 impressions
Balanced accuracy: 0.5055

              precision    recall  f1-score   support

           0      0.837     0.979     0.902     13939
           1      0.232     0.032     0.056      2751

    accuracy                          0.823     16690
   macro avg      0.534     0.505     0.479     16690
weighted avg      0.737     0.823     0.763     16690



In [34]:
print(data.columns.tolist())

['client_hash_id', 'content_hash_id', 'imp_prev90', 'clk_prev90', 'ctr_prev90', 'position_volatility', 'weekend_share', 'position_recent45', 'position_older45', 'position_change', 'imp_future30', 'is_declining']


In [35]:
grain_check = con.sql(f"""
    SELECT
        COUNT(*) AS total_rows,
        COUNT(DISTINCT client_hash_id || '|' || content_hash_id || '|' || CAST(report_date AS VARCHAR))
            AS distinct_client_content_date,
        COUNT(*) -
        COUNT(DISTINCT client_hash_id || '|' || content_hash_id || '|' || CAST(report_date AS VARCHAR))
            AS duplicate_rows
    FROM {TABLES['fact_daily']}
    WHERE report_date >= DATE '2026-03-01'
      AND report_date < DATE '2026-04-01'
""").df()

grain_check

FloatProgress(value=0.0, layout=Layout(width='auto'), style=ProgressStyle(bar_color='black'))

,total_rows,distinct_client_content_date,duplicate_rows
0,9841378,9841378,0


In [36]:
date_span_check = con.sql(f"""
    SELECT
        COUNT(*) AS row_count,
        MIN(report_date) AS first_date,
        MAX(report_date) AS last_date
    FROM {TABLES['fact_daily']}
    WHERE report_date >= DATE '2026-03-01'
      AND report_date < DATE '2026-04-01'
""").df()

date_span_check

FloatProgress(value=0.0, layout=Layout(width='auto'), style=ProgressStyle(bar_color='black'))

,row_count,first_date,last_date
0,9841378,2026-03-01,2026-03-31


In [37]:
con.sql(f"""
    DESCRIBE
    SELECT *
    FROM {TABLES['fact_daily']}
    LIMIT 1
""").df()

,column_name,column_type,null,key,default,extra
0,report_date,DATE,YES,None,None,None
1,client_hash_id,VARCHAR,YES,None,None,None
2,content_hash_id,VARCHAR,YES,None,None,None
3,client_has_gsc,BOOLEAN,YES,None,None,None
4,client_has_ga4,BOOLEAN,YES,None,None,None
5,gsc_data_available,BOOLEAN,YES,None,None,None
6,ga4_data_available,BOOLEAN,YES,None,None,None
7,gsc_impressions,BIGINT,YES,None,None,None
8,gsc_clicks,BIGINT,YES,None,None,None
9,gsc_sum_position,BIGINT,YES,None,None,None


In [38]:
availability_check = con.sql(f"""
    SELECT
        COUNT(*) AS march_rows,
        COUNT(*) FILTER (WHERE gsc_data_available IS TRUE) AS available_rows
    FROM {TABLES['fact_daily']}
    WHERE report_date >= DATE '2026-03-01'
      AND report_date < DATE '2026-04-01'
""").df()

availability_check

FloatProgress(value=0.0, layout=Layout(width='auto'), style=ProgressStyle(bar_color='black'))

,march_rows,available_rows
0,9841378,3611061


In [39]:
con.sql(f"""
    DESCRIBE
    SELECT *
    FROM {TABLES['fact_query_90d']}
    LIMIT 1
""").df()

,column_name,column_type,null,key,default,extra
0,client_hash_id,VARCHAR,YES,None,None,None
1,content_hash_id,VARCHAR,YES,None,None,None
2,query_hash_id,VARCHAR,YES,None,None,None
3,query_char_count,BIGINT,YES,None,None,None
4,query_token_count,BIGINT,YES,None,None,None
5,window_start,DATE,YES,None,None,None
6,window_end,DATE,YES,None,None,None
7,impressions_90d,BIGINT,YES,None,None,None
8,clicks_90d,BIGINT,YES,None,None,None
9,impressions_last30,BIGINT,YES,None,None,None


In [40]:
features = con.sql(f"""
    SELECT
        client_hash_id,
        content_hash_id,

        SUM(gsc_impressions) AS imp_prev45,

        SUM(gsc_clicks) AS clk_prev45,

        SUM(gsc_clicks) / NULLIF(SUM(gsc_impressions), 0)
            AS ctr_prev45,

        STDDEV(gsc_avg_position) AS position_volatility,

        AVG(
            CASE
                WHEN DAYOFWEEK(report_date) IN (0, 6)
                THEN 1.0
                ELSE 0.0
            END
        ) AS weekend_share

    FROM {TABLES['fact_daily']}

    WHERE report_date >= DATE '2026-01-15'
      AND report_date < DATE '2026-03-01'
      AND gsc_data_available IS TRUE

    GROUP BY
        client_hash_id,
        content_hash_id

    HAVING SUM(gsc_impressions) >= 150
""").df()

print(f"Feature rows: {len(features):,}")
features.head()

FloatProgress(value=0.0, layout=Layout(width='auto'), style=ProgressStyle(bar_color='black'))

Feature rows: 79,428


,client_hash_id,content_hash_id,imp_prev45,clk_prev45,ctr_prev45,position_volatility,weekend_share
0,client_23a62021009f63c4,content_17e1f9590767a167,11764.0,5.0,0.000425,17.927764,0.288889
1,client_23a62021009f63c4,content_0624639f874d5630,12766.0,88.0,0.006893,2.088690,0.288889
2,client_23a62021009f63c4,content_e8be016907340387,7497.0,15.0,0.002001,5.445825,0.288889
3,client_23a62021009f63c4,content_e4bdcbda48c7ce40,2117.0,19.0,0.008975,5.832424,0.288889
4,client_23a62021009f63c4,content_7692b4b1ac62ada8,1915.0,1.0,0.000522,5.795256,0.288889


### Five features and availability

| Feature | Available when? |
|---|---|
| `imp_prev45` | Knowable at the decision moment because it is calculated only from GSC impressions in the preceding 45-day historical window. |
| `clk_prev45` | Knowable at the decision moment because it is calculated only from historical GSC clicks in the preceding 45-day window. |
| `ctr_prev45` | Knowable at the decision moment because it is calculated from historical clicks and impressions before the decision moment. |
| `position_volatility` | Knowable at the decision moment because it measures variation in historical average search position before the decision moment. |
| `weekend_share` | Knowable at the decision moment because it is calculated from the historical distribution of observations across weekdays and weekends. |

In [41]:
query_dates = con.sql(f"""
    SELECT
        MIN(window_start) AS earliest_window_start,
        MAX(window_end) AS latest_window_end,
        COUNT(*) AS total_rows
    FROM {TABLES['fact_query_90d']}
""").df()

query_dates

FloatProgress(value=0.0, layout=Layout(width='auto'), style=ProgressStyle(bar_color='black'))

,earliest_window_start,latest_window_end,total_rows
0,2026-04-02,2026-06-30,2414248


In [42]:
features.info()
features.isna().sum()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 79428 entries, 0 to 79427
Data columns (total 7 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   client_hash_id       79428 non-null  object 
 1   content_hash_id      79428 non-null  object 
 2   imp_prev45           79428 non-null  float64
 3   clk_prev45           79428 non-null  float64
 4   ctr_prev45           79428 non-null  float64
 5   position_volatility  79427 non-null  float64
 6   weekend_share        79428 non-null  float64
dtypes: float64(5), object(2)
memory usage: 4.2+ MB


,0
client_hash_id,0
content_hash_id,0
imp_prev45,0
clk_prev45,0
ctr_prev45,0
position_volatility,1
weekend_share,0


In [43]:
future_outcome = con.sql(f"""
    SELECT
        client_hash_id,
        content_hash_id,
        SUM(gsc_impressions) AS imp_future30

    FROM {TABLES['fact_daily']}

    WHERE report_date >= DATE '2026-03-01'
      AND report_date < DATE '2026-04-01'
      AND gsc_data_available IS TRUE

    GROUP BY
        client_hash_id,
        content_hash_id
""").df()

print(f"Future outcome rows: {len(future_outcome):,}")
future_outcome.head()

data = features.merge(
    future_outcome,
    on=['client_hash_id', 'content_hash_id'],
    how='inner'
)

data['is_declining'] = (
    (data['imp_future30'] / 30)
    < 0.8 * (data['imp_prev45'] / 45)
).astype(int)

print("Model rows:", len(data))
print("Declining rate:", data['is_declining'].mean())

FloatProgress(value=0.0, layout=Layout(width='auto'), style=ProgressStyle(bar_color='black'))

Future outcome rows: 176,738
Model rows: 75460
Declining rate: 0.18869599787967134


In [44]:
from sklearn.model_selection import GroupShuffleSplit
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import balanced_accuracy_score

feature_cols = [
    'imp_prev45',
    'clk_prev45',
    'ctr_prev45',
    'position_volatility',
    'weekend_share'
]

model_data = data.dropna(
    subset=feature_cols + ['is_declining']
).copy()

X = model_data[feature_cols]
y = model_data['is_declining']
groups = model_data['client_hash_id']

splitter = GroupShuffleSplit(
    n_splits=1,
    test_size=0.25,
    random_state=42
)

train_idx, test_idx = next(splitter.split(X, y, groups=groups))

X_train = X.iloc[train_idx]
X_test = X.iloc[test_idx]
y_train = y.iloc[train_idx]
y_test = y.iloc[test_idx]

model = RandomForestClassifier(
    n_estimators=200,
    random_state=42,
    n_jobs=-1
)

model.fit(X_train, y_train)

pred = model.predict(X_test)

honest_score = balanced_accuracy_score(y_test, pred)

print(f"Honest balanced accuracy: {honest_score:.4f}")

Honest balanced accuracy: 0.5122


In [45]:
leaky_cols = feature_cols + ['imp_future30']

X_leaky = model_data[leaky_cols]

X_train_l = X_leaky.iloc[train_idx]
X_test_l = X_leaky.iloc[test_idx]

leaky_model = RandomForestClassifier(
    n_estimators=200,
    random_state=42,
    n_jobs=-1
)

leaky_model.fit(X_train_l, y_train)

leaky_pred = leaky_model.predict(X_test_l)

leaky_score = balanced_accuracy_score(y_test, leaky_pred)

print(f"Leaky balanced accuracy: {leaky_score:.4f}")

Leaky balanced accuracy: 0.9789


In [46]:
# Summarize the Week 3 experiments

results = pd.DataFrame({
    "experiment": [
        "90-day, >=300 impressions, no position_change",
        "90-day, >=300 impressions, with position_change",
        "90-day, >=500 impressions, with position_change"
    ],
    "model_rows": [
        len(model_data),
        len(model_data),
        len(model_data_500)
    ],
    "declining_rate": [
        y_base.mean(),
        y.mean(),
        y_500.mean()
    ],
    "balanced_accuracy": [
        balanced_accuracy_score(y_test_base, baseline_pred),
        balanced_accuracy_score(y_test, pred),
        balanced_accuracy_score(y_test_500, pred_500)
    ]
})

print(results.to_string(index=False))

                                     experiment  model_rows  declining_rate  balanced_accuracy
  90-day, >=300 impressions, no position_change       75459        0.200317           0.521815
90-day, >=300 impressions, with position_change       75459        0.188698           0.512249
90-day, >=500 impressions, with position_change       51220        0.199043           0.505482


## Week 3 findings

The 90-day historical feature experiment tested whether recent search-performance signals could help identify content pages that would experience a substantial decline in impressions during March 2026.

Three experiments were compared:

- 90-day history with a minimum of 300 impressions and no `position_change`
- 90-day history with a minimum of 300 impressions and `position_change`
- 90-day history with a minimum of 500 impressions and `position_change`

All models used a Random Forest and `GroupShuffleSplit` by `client_hash_id` to prevent the same client from appearing in both the training and test sets.

The results showed limited predictive power. Balanced accuracy was 0.5218 without `position_change`, 0.5122 after adding `position_change`, and 0.5055 after increasing the minimum impression threshold to 500.

Adding `position_change` therefore did not improve performance in this experiment. Increasing the minimum historical impression threshold also did not improve performance.

The results should be interpreted as an experiment on this particular dataset, feature set, target definition, and time period rather than evidence that these signals can never be useful. The low balanced accuracy indicates that additional features, alternative target definitions, or different modeling approaches would be needed for stronger predictive performance.

The experiment also demonstrates the importance of evaluating imbalanced classification problems with metrics such as balanced accuracy rather than relying only on overall accuracy.

## 4) Deliberate leakage check

I deliberately added `imp_future30` as a feature even though it is derived
from the future outcome window.

| Model | Balanced accuracy |
|---|---:|
| Honest model | 0.5132 |
| Leaky model | 0.9778 |

The leaky model's score increased dramatically from 0.5132 to 0.9778.
This confirms that `imp_future30` contains information about the target
that would not be available at the decision moment.

`imp_future30` was therefore removed from the feature set. The retained
honest balanced accuracy is **0.5132**.

### Named limitation

The current historical features have limited predictive power for the
decline label, as shown by the honest balanced accuracy of 0.5132.
In addition, the query-level warehouse table begins on 2026-04-02, so
query-level features cannot be used for the March 2026 development slice
without creating a temporal mismatch.

In [48]:
# Week 3 final sanity check

final_features = [
    'imp_prev90',
    'clk_prev90',
    'ctr_prev90',
    'position_volatility',
    'weekend_share',
    'position_change'
]

print("=== WEEK 3 SANITY CHECK ===")
print("Warehouse connection: OK")
print(f"Final feature rows: {len(model_data):,}")
print(f"Features used: {final_features}")
print("Target: is_declining")
print("Model: Random Forest")
print("Split: GroupShuffleSplit by client_hash_id")
print(f"Balanced accuracy: {balanced_accuracy_score(y_test, pred):.4f}")
print("Status: Week 3 experiment complete")

=== WEEK 3 SANITY CHECK ===
Warehouse connection: OK
Final feature rows: 75,459
Features used: ['imp_prev90', 'clk_prev90', 'ctr_prev90', 'position_volatility', 'weekend_share', 'position_change']
Target: is_declining
Model: Random Forest
Split: GroupShuffleSplit by client_hash_id
Balanced accuracy: 0.5122
Status: Week 3 experiment complete


Whatever number you just got: interrogate it before you believe it. Which feature carries the
signal? Does it survive a per-client split (train on some clients, test on others)? That
question — *does it generalize across clients?* — is exactly what separates a capstone-grade
result from a lucky split.

## Your turn

1. Re-run section 3 with a **90-day** window and a `HAVING` threshold of your choice.
2. Add one feature you believe in (position volatility? weekend share? query concentration?).
3. Replace the random split with **GroupShuffleSplit on `client_hash_id`** and compare.

## Working locally instead

```python
from huggingface_hub import snapshot_download
path = snapshot_download(repo_id='FlyRank/internship-warehouse', repo_type='dataset',
                         allow_patterns=['dim_*.parquet', 'fact_content_query_90d.parquet',
                                         'fact_content_daily_performance/month=2026-0*/*.parquet'])
```
Then point `REL` at that local path. Download only the month partitions you need — the
`allow_patterns` filter above is the whole trick.

---

**Where this fits:** every lane brief assumes you can produce per-content feature tables like
the one you just built. The lane datasets under the `lanes` HF repo are pre-cut examples of
exactly this pattern — but for the capstone, features you engineered yourself from the full
release beat any pre-cut file.
